src/operator/numpy/np_indexing_op.cc

```diff
@@ -0,0 +1,572 @@
+      is_valid = false;
+      break;
+    }
+  }
+  return is_valid;
+}
+
+inline bool AdvancedIndexingOpType(const nnvm::NodeAttrs& attrs,
+                       std::vector<int> *in_attrs,
+                       std::vector<int> *out_attrs) {
+  CHECK_EQ(in_attrs->size(), 2U);
+  CHECK_EQ(out_attrs->size(), 1U);
+  CHECK_NE((*in_attrs)[1], -1) << "Index type must be set for take operator";
+
+  TYPE_ASSIGN_CHECK(*out_attrs, 0, (*in_attrs)[0]);
+  TYPE_ASSIGN_CHECK(*in_attrs, 0, (*out_attrs)[0]);
+  return (*in_attrs)[0] != -1;
+}
+
+bool AdvancedIndexingOpStorageType(const nnvm::NodeAttrs& attrs,
+                            const int dev_mask,
+                            DispatchMode* dispatch_mode,
+                            std::vector<int> *in_attrs,
+                            std::vector<int> *out_attrs) {
+  CHECK_EQ(in_attrs->size(), 2);
+  CHECK_EQ(out_attrs->size(), 1);
+  for (int &attr : *in_attrs) {
+    CHECK_EQ(attr, kDefaultStorage) << "Only default storage is supported";
+  }
+  for (int &attr : *out_attrs) {
+    attr = kDefaultStorage;
+  }
+  *dispatch_mode = DispatchMode::kFComputeEx;
+  return true;
+}
+
+bool AdvancedIndexingOpBackStorageType(const nnvm::NodeAttrs& attrs,
+                                const int dev_mask,
+                                DispatchMode* dispatch_mode,
+                                std::vector<int> *in_attrs,
+                                std::vector<int> *out_attrs) {
+  CHECK_EQ(in_attrs->size(), 3);
+  CHECK_EQ(out_attrs->size(), 2);
+  for (int &attr : *in_attrs) {
+    CHECK_EQ(attr, kDefaultStorage) << "Only default storage is supported";
+  }
+  for (int &attr : *out_attrs) {
+    attr = kDefaultStorage;
+  }
+  for (size_t i = 0; i < out_attrs->size(); i++)
+    out_attrs->at(i) = kDefaultStorage;
+  *dispatch_mode = DispatchMode::kFComputeEx;
+  return true;
+}
+
+template<>
+void AdvancedIndexingOpForward<cpu>(const nnvm::NodeAttrs& attrs,
+                        const OpContext& ctx,
+                        const std::vector<NDArray>& inputs,
+                        const std::vector<OpReqType>& req,
+                        const std::vector<NDArray>& outputs) {
+  using namespace mxnet_op;
+  if (req[np_indexing_::kOut] == kNullOp) return;
+  CHECK_EQ(inputs.size(), 2U);
+  CHECK_EQ(outputs.size(), 1U);
+
+  if (inputs[np_indexing_::kIdx].dtype() == mshadow::kBool) {
+    CHECK(req[0] == kWriteTo || req[0] == kWriteInplace);
+    const NDArray &data = inputs[0];
+    const NDArray &idx = inputs[1];
+    const NDArray &out = outputs[0];
+    CHECK_EQ(data.shape()[0], idx.shape()[0]);
+    CHECK_EQ(idx.shape().ndim(), 1U);  // idx is required to be 1-d.
+    // count the number of 1s in `idx`, so that we could know the output dimension
+    size_t idx_size = idx.shape()[0];
+    std::vector<int32_t> prefix_sum(idx_size, 0);
+    size_t valid_num = 0;
+    // Calculate prefix sum
+    MSHADOW_TYPE_SWITCH_WITH_BOOL(idx.dtype(), DType, {
```

Comment:
There is no need to use `MSHADOW_TYPE_SWITCH_WITH_BOOL` since we have already tested that it's a bool type.